tempesta_fw/http_sticky.c

```diff
@@ -0,0 +1,740 @@
+{
+	char *buf;
+	size_t len;
+	TfwStr *hdr;
+
+	if ((hdr = tfw_http_field(hm, field->ptr, field->len)) == NULL) {
+		return 0;
+	}
+	len = tfw_str_len(hdr) + 1;
+	if ((buf = tfw_pool_alloc(hm->pool, len)) == NULL) {
+		return -ENOMEM;
+	}
+	len = tfw_str_to_cstr(hdr, buf, len);
+	value->ptr = buf + field->len;
+	value->len = len - field->len;
+	tfw_http_strim(value);
+
+	return 1;
+}
+
+/*
+ * Find Tempesta sticky cookie in an HTTP message.
+ */
+static int
+tfw_http_sticky_get(TfwHttpMsg *hm, tfw_str_t *cookie)
+{
+	int ret;
+	u_char *valptr, *endptr;
+	const tfw_str_t s_field = tfw_string("Cookie:");
+	tfw_str_t value = { 0 };
+
+	/*
+	 * Find a 'Cookie:' header field in the request.
+	 * The search for Tempesta sticky cookie within the field.
+	 * Note that there can be only one "Cookie:" header field.
+	 */
+	if ((ret = tfw_http_field_val(hm, &s_field, &value)) <= 0) {
+		return ret;
+	}
+	valptr = strnstr(value.ptr, tfw_cfg_sticky.name.ptr, value.len);
+	if (!valptr)
+		return 0;
+	cookie->ptr = valptr + tfw_cfg_sticky.name.len + 1;
+
+	valptr = cookie->ptr;
+	endptr = value.ptr + value.len;
+	while((valptr < endptr) && (*valptr != ';') && !isspace(*valptr))
+		valptr++;
+	cookie->len = valptr - cookie->ptr;
+
+	return 1;
+}
+
+/*
+ * Set up complete 'Set-Cookie:' header field for a connection.
+ */
+	tfw_str_t value = {
+		.len = sizeof("test") - 1,
+		.ptr = "test"
+	};
+/*
+ * Create Tempesta sticky cookie value and set it for the client.
+ *
+ * Tempesta sticky cookie is based on:
+ * - HTTP request source IP address;
+ * - HTTP request User-Agent string;
+ * - Current timestamp;
+ */
+static int
+tfw_http_sticky_set(TfwHttpMsg *hm)
+{
+	int ret;
+	uint64_t nsecs;
+	struct timespec ts;
+	tfw_str_t ua_value = { 0 };
+	const tfw_str_t s_field = tfw_string("User-Agent:");
+	TfwClient *client = (TfwClient *)hm->conn->peer;
+
+	char desc[sizeof(struct shash_desc)
+		  + crypto_shash_descsize(tfw_sticky_shash)]
+		  CRYPTO_MINALIGN_ATTR;
+	struct shash_desc *shash_desc = (struct shash_desc *)desc;
+
+	memset(desc, 0, sizeof(desc));
+	shash_desc->tfm = tfw_sticky_shash;
+	shash_desc->flags = 0;
+
+	if ((ret = tfw_http_field_val(hm, &s_field, &ua_value)) <= 0) {
+		return ret;
+	}
+
+	getnstimeofday(&ts);
+	nsecs = 1000ULL * 1000 * 1000 * ts.tv_sec + ts.tv_nsec;
+
+//	client->sock->sk_socket
+
+	crypto_shash_init(shash_desc);
+	crypto_shash_update(shash_desc, (u8 *)ua_value.ptr, ua_value.len);
+	crypto_shash_finup(shash_desc, (u8 *)&nsecs, sizeof(nsecs), client->hmac);
+
+	return 0;
+}
+
+/*
+ * Add Tempesta sticky cookie to an HTTP response.
+ *
+ * Create a complete 'Set-Cookie:' header field, and add it
+ * to the HTTP response' header block.
+ */
+static int
+tfw_http_sticky_add(TfwHttpMsg *hm, u_char *value, size_t len)
+{
+	int ret;
+	char buf[256], *ptr = buf;
+
+	/* Set up complete 'Set-Cookie:' header field */
+	memcpy(ptr, S_F_SET_COOKIE, SLEN(S_F_SET_COOKIE));
```

Comment:
You don't care about data copies at all. At least for this case you could initialize buf with the static string to avoid at least this memcpy().